### DISCUSSION TITLE
Change Effect Config Parameter during runtime

### DISCUSSION THREAD

**Discussion by achronoxx** [PlummersSoftwareLLC/NightDriverStrip#615] [Q&A]
I'd like to finetune some of the built-in effects while they are running, so I can get a better understanding of what the various parameters in an effects config do.  For example, EFFECT_STRIP_FIRE has the following configurable parameters:
FireEffect(
const String & strName, 
int ledCount = NUM_LEDS, 
int cellsPerLED = 1, 
int cooling = 20, 
int sparking = 100, 
int sparks = 3, 
int sparkHeight = 4, 
bool breversed = false, 
bool bmirrored = false)

I was successful in using the Postman App to send the REST_API commands, and in particular, the "/effectsConfig" command gave me the list of all the parameters as follows:

{
    "ver": 1,
    "ivl": 30000,
    "prj": "Demo",
    "esv": 1,
    "efs": [
        {
            "en": 16,
            "fn": "RainbowFill Rainbow",
            "es": 1,
            "mi": 1,
            "ce": 1,
            "sdd": 6,
            "dth": 2
        },
        {
            "en": 18,
            "fn": "Status Fill",
            "es": 1,
            "mi": 1,
            "ce": 1,
            "ent": 10,
            "clr": 32768
        },
        {
            "en": 1,
            "fn": "Bouncing Balls",
            "es": 1,
            "mi": 1,
            "ce": 1,
            "blc": 3,
            "bls": 1,
            "mir": true,
            "ers": true
        },
        {
            "en": 8,
            "fn": "Medium Fire",
            "es": 1,
            "mi": 1,
            "ce": 1,
            "ldc": 72,
            "cpl": 1,
            "clg": 3,
            "spc": 3,
            "sph": 4,
            "spg": 100,
            "rvr": true,
            "mir": true
        },

This is exactly what I was looking for, but I don't know how to change this information.  For example, I want to change "spc" from 3 to 5.

I hope someone can help me figure this out.
   
Thanks,
Andrew

**Comment by rbergen** [comment]
It is possible to change the settings of an effect "at runtime", but only if and to the extent the effect in question explicitly allows it. The reason for this is that most effects are now set up to only accept parameters when they are constructed, and just changing those later on can easily cause problems if the effect  has not been developed to deal with such changes.

If you open the web UI on your device, you can click on the settings cog wheel for an effect to see which settings the effect exposes for real-time modification. Some of them (name, maximum effect time) are exposed by the effect base class that is LEDStripEffect. Individual effects can add their own properties to that. If I recall correctly, one effect that does that currently is PatternSubscribers. That's an effect for the Mesmerizer configuration, but the way the class exposes and processes its additional settings applies in general. 

There's actually a long-open feature request in the issue list (#294) that invites contributors to extend the current set of effect-specific settings.

**Reply by achronoxx** [reply]
Hi Rutger,

Glad to see this is captured as an issue.  I quickly read over issue #294 and it seems the proposed solution is to make the effect-changes quickly in 'realtime'.  This would be ideal, but I don't think we necessarily want quick-realtime changes.  I'm just trying to avoid rebuilding/uploading the project when I make a simple effect change.  I'd like to suggest another alternative that may be simpler. 

- Update the REST_API to allow you to change the effects parameters in the JSON (eg. change "spc" parameter)
- Use the existing Web UI to simply restart the effect, which will force it to re-read the updated JSON file (I assume the JSON file is read each time an effect starts, please correct me if I'm wrong about this)

So, this is just one change to allow us to modify the "/effectsConfig" JSON file.

Also, where are the JSON files (SPIFFS?) stored - in FLASH, EEPROM or RAM?

Please let me know what you think of this idea.

Thanks,
Andrew

**Reply by rbergen** [reply]
I understand the request, but I'm not eager to make this part of the main tree (and yes, that's an understatement). The reason is that making *any* mistake in the composition of the JSON you can now request via the /effectsConfig endpoint is a fail-safe way to make sure _no_ effects load on device startup. That will then remain the situation until you do a full effects config reset. It just doesn't feel responsible to make that behavior a "supported" scenario for the general user. Of course, that decision on my part does not block you from adding the "unfiltered write" capability in a branch in your own fork, to support any local testing you'd like.

The /effectsConfig endpoint is an endpoint (implemented by the CWebServer class), not a file as such. The actual file is called "/effects.cfg", which is stored on the SPIFFS file system in the relevant partition on flash.

That file is read once on device startup (_not_ whenever an effect is started), and then written at runtime whenever something happens that warrants an update to it - an update to an effect setting being a prime example of such an event.

Cheers,
Rutger

**Reply by achronoxx** [reply]
Hi Rutger,

Thanks for your thoughtful response - I totally understand your point.

I don't have any experience with SPIFF files, but I may try making this change in my private branch as a learning opportunity!

Andrew

**Comment by robertlipe** [comment]
Yes, that's the water I'm pretty sure he was leading you to. :-) Sometimes
we all want crazy things that just don't belong in the main tree for
various reasons - then it's up to us to create and maintain such things.
That's one of the points of open source. (In another group right now, Im
watching someone posit that "open source" means an endless license for the
maintainers to work for them for free. It's not going to end well for him.)

Don't get hung up on SPIFFS. That's just the filesystem. You don't think
about NTFS, FAT,16, VXFS, HTFS or whatever file your computer uses. You
just edit or copy files. How the files are stored on the floppy/flash
chip/disk drive/abacus is not really relevant. (If it were relevant to you,
you'd already know this. It's like how time "just works" - unless you're
the one managing timer interrupts, rescheduling tasks, managing locks, and
all the other hard work to make it look easy.)
